libp2p/stream_muxer/muxer_multistream.py

```diff
@@ -0,0 +1,70 @@
+from typing import Dict, List, Mapping, Type
+
+from libp2p.network.connection.raw_connection_interface import IRawConnection
+from libp2p.network.typing import GenericProtocolHandlerFn
+from libp2p.peer.id import ID
+from libp2p.protocol_muxer.multiselect import Multiselect
+from libp2p.protocol_muxer.multiselect_client import MultiselectClient
+from libp2p.protocol_muxer.multiselect_communicator import RawConnectionCommunicator
+from libp2p.security.secure_conn_interface import ISecureConn
+from libp2p.typing import TProtocol
+
+from .abc import IMuxedConn
+
+MuxerClassType = Type[IMuxedConn]
+
+# FIXME: add negotiate timeout to `MuxerMultistream`
+DEFAULT_NEGOTIATE_TIMEOUT = 60
+
+
+class MuxerMultistream:
+    """
+    MuxerMultistream is a multistream stream muxed transport multiplexer.
+    go implementation: github.com/libp2p/go-stream-muxer-multistream/multistream.go
+    """
+
+    transports: Dict[TProtocol, MuxerClassType]
+    multiselect: Multiselect
+    multiselect_client: MultiselectClient
+    order_preference: List[TProtocol]
+
+    def __init__(
+        self, muxer_transports_by_protocol: Mapping[TProtocol, MuxerClassType]
+    ) -> None:
+        self.transports = {}
+        self.multiselect = Multiselect()
+        self.multiselect_client = MultiselectClient()
+        self.order_preference = []
+        for protocol, transport in muxer_transports_by_protocol.items():
+            self.add_transport(protocol, transport)
+
+    def add_transport(self, protocol: TProtocol, transport: MuxerClassType) -> None:
+        self.transports[protocol] = transport
```

Comment:
Should there be a check here to see if the `protocol/transport` is already present?  What should the behavior be if it is?  Should the `protocol` instance be removed from `order_preference` and appended to the end?  Otherwise it will be in the list twice.

Reply:
> Should there be a check here to see if the protocol/transport is already present? 

It is not checked in Go. It just removes the previous one and adds it again. Reference: https://github.com/libp2p/go-stream-muxer-multistream/blob/master/multistream.go#L35-L39

> Should the protocol instance be removed from order_preference and appended to the end?

Good catch. It should be but it is not handled in Go. I will change it.

Reply:
@pipermerriam 

I turn out to use `OrderedDict` for `MuxerMultistream.transport`. This way we can just use `transport.keys()` as the precedence when negotiation. I also remove a protocol in `add_transport` when it was already added before, to ensure that its key is reordered.